xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+// xtext.cpp
+// Lua interface to wrap a utf8 string into a XText object
+// that provides various text shaping and layout methods
+// with the help of Fribidi, Harfbuzz and libunibreak.
+
+// We do many things similarly to how they are done in crengine,
+// and took and adapted much code from it.
+// For many links and notes about the concepts and libraries used,
+// see: https://github.com/koreader/crengine/issues/307
+
+extern "C"
+{
+#include <lua.h>
+#include <lauxlib.h>
+#include <lualib.h>
+#include "xtext.h"
+}
+
+// Harfbuzz
+#include <hb.h>
+#include <hb-ft.h>
+
+// FriBiDi
+#include <fribidi/fribidi.h>
+
+// libunibreak
+#include <wordbreak.h>
+#include <linebreak.h>
+    // linebreakdef.h is not wrapped by this, unlike linebreak.h
+    // (not wrapping it results in "undefined symbol" with the
+    // original function name kinda obfuscated)
+    #ifdef __cplusplus
+    extern "C" {
+    #endif
+#include <linebreakdef.h>
+    #ifdef __cplusplus
+    }
+    #endif
+
+// luajit doesn't provide lua_set/getuservalue (unless compiled
+// with LUAJIT_ENABLE_LUA52COMPAT) but these are equivalents
```

Comment:
Well, there's probably no real harm in enabling that. Will LUA_VERSION_NUM be 502 if that's enabled?

Reply:
Dunno, but not taking any risk. That's mostly just some rename from the Lua5.2 people who wanted to get rid of the word "environment".
Really no need to take a risk. It enables quite a few more risk/incompatible features.

Reply:
I believe I compiled with that and it worked fine; if you check my pull requests in front there was one tiny fix I committed after testing it.

Reply:
Anyway, I guess I'm asking if it's future-compatible with enabling that flag.

Reply:
I guess this PR would work with that flag enabled. But haven't checked what LUA_VERSION_NUM would be. Anyway, it would be a simple fix of removing these #define.
But if you'd like to enable it, better do it later, after all that RTL UI/text/keyboard stuff is done, to not add uneeded noise and errors :)

Reply:
Oh, here it is: koreader/koreader#3090

I guess I left it somewhat implied, or maybe it was also related to a newer LuaJIT version tout court? I don't recall the exact specifics, it's been a little while.